cloud/amazon/ec2_search.py

```diff
@@ -0,0 +1,157 @@
+    import boto
+    import boto.ec2
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+
+def todict(obj, classkey=None):
+    if isinstance(obj, dict):
+        data = {}
+        for (k, v) in obj.items():
+            data[k] = todict(v, classkey)
+        return data
+    elif hasattr(obj, "_ast"):
+        return todict(obj._ast())
+    elif hasattr(obj, "__iter__"):
+        return [todict(v, classkey) for v in obj]
+    elif hasattr(obj, "__dict__"):
+        # This Class causes a recursive loop and at this time is not worth
+        # debugging. If it's useful later I'll look into it.
+        if not isinstance(obj, boto.ec2.blockdevicemapping.BlockDeviceType):
+            data = dict([(key, todict(value, classkey))
+                for key, value in obj.__dict__.iteritems()
+                if not callable(value) and not key.startswith('_')])
+            if classkey is not None and hasattr(obj, "__class__"):
+                data[classkey] = obj.__class__.__name__
+            return data
+    else:
+        return obj
+
+def get_all_ec2_regions(module):
+    try:
+        regions = boto.ec2.regions()
+    except Exception, e:
+        module.fail_json('Boto authentication issue: %s' % e)
+
+    return regions
+
+# Connect to ec2 region
+def connect_to_region(region, module):
+    try:
+        conn = boto.ec2.connect_to_region(region.name)
+    except Exception, e:
+        print module.jsonify('error connecting to region: ' + region.name)
+        conn = None
+    # connect_to_region will fail "silently" by returning
+    # None if the region name is wrong or not supported
+    return conn
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            key = dict(default='Name'),
+            value = dict(),
+            lookup = dict(default='tags'),
+            ignore_state = dict(default='terminated'),
+            region = dict(),
+        )
+    )
+
+    if not HAS_BOTO:
+        module.fail_json(msg='boto required for this module')
+
+    server_info = list()
+
+    for region in get_all_ec2_regions(module):
+        conn = connect_to_region(region, module)
+        try:
+            # Run when looking up by tag names, only returning hostname currently
+            if module.params.get('lookup') == 'tags':
+                ec2_key = 'tag:' + module.params.get('key')
+                ec2_value = module.params.get('value')
+                reservations = conn.get_all_instances(filters={ec2_key : ec2_value})
+                for instance in [i for r in reservations for i in r.instances]:
+                    if instance.private_ip_address != None:
+                        instance.hostname = 'ip-' + instance.private_ip_address.replace('.', '-')
+                    if instance._state.name not in module.params.get('ignore_state'):
+                        server_info.append(todict(instance))
+        except:
+            print module.jsonify('error getting instances from: ' + region.name)
```

Comment:
wont fail json cause the script to exit? it will always fail on the us-gov region. would it be better to add a param to filter out this region since very few people will use it but it is still possible that government contractors use ansible.